libsolidity/interface/CompilerStack.cpp

```diff
@@ -180,7 +181,27 @@ bool CompilerStack::parse()
 				if (!resolver.updateDeclaration(*m_globalContext->currentThis())) return false;
 				if (!resolver.updateDeclaration(*m_globalContext->currentSuper())) return false;
 				if (!resolver.resolveNamesAndTypes(*contract)) return false;
-				m_contracts[contract->name()].contract = contract;
+
+				if (m_contracts.find(contract->fullyQualifiedName()) != m_contracts.end())
```

Comment:
I noticed the double-reporting of the error.  Actually, it's worth noting that the only way to cause a collision now is to have the same contract twice in one file, which is actually covered under another error already, anyway.  So, I think you're right that it's actually okay to simply silently continue without adding the contract since it's covered by the other error.

I will, however, be leaving comments explaining what's going on in the code, because it feels like the sort of thing that a future developer could trip over.